lib/ansible/plugins/callback/grafana_annotations.py

```diff
@@ -0,0 +1,203 @@
+    description:
+      - This callback will report start, failed and stats events to Grafana as annotations (https://grafana.com)
+    version_added: "2.5"
+    requirements:
+      - whitelisting in configuration
+    options:
+      server:
+        description: Address of the Logstash server
+        env:
+          - name: GRAFANA_SERVER
+        default: localhost
+      port:
+        description: Port on which Grafana is listening
+        env:
+            - name: GRAFANA_PORT
+        default: 3000
+      secure:
+        description: Boolean specifying if HTTPS should be used. (0=HTTP, 1=HTTPS)
+        env:
+          - name: GRAFANA_SECURE
+        default: 0
+      token:
+        description: Grafana API Token, allowing to authenticate when posting on the HTTP API.
+                     If not provided, the callback will be disabled.
+        env:
+          - name: GRAFANA_API_TOKEN
+'''
+
+import os
+import json
+import socket
+import getpass
+from datetime import datetime
+
+try:
+    import httplib
+except ImportError:
+    # Python 3
+    import http.client as httplib
+
+from ansible.plugins.callback import CallbackBase
+
+
+PLAYBOOK_START_TXT = """\
+Started playbook {playbook}
+
+From '{hostname}'
+By user '{username}'
+"""
+
+PLAYBOOK_ERROR_TXT = """\
+Playbook {playbook} Failure !
+
+From '{hostname}'
+By user '{username}'
+
+'{task}' failed on {host}
+
+debug: {result}
+"""
+
+PLAYBOOK_STATS_TXT = """\
+Playbook {playbook}
+Duration: {duration}
+Status: {status}
+
+From '{hostname}'
+By user '{username}'
+
+Result:
+{summary}
+"""
+
+
+def to_millis(dt):
+    return int(dt.strftime('%s')) * 1000
+
+
+class CallbackModule(CallbackBase):
+    """
+    ansible grafana callback plugin
+    ansible.cfg:
+        callback_plugins   = <path_to_callback_plugins_folder>
+        callback_whitelist = grafana
+    and put the plugin in <path_to_callback_plugins_folder>
+
+    This plugin makes use of the following environment variables:
+        GRAFANA_SERVER   (optional): defaults to localhost
+        GRAFANA_PORT     (optional): defaults to 3000
+        GRAFANA_SECURE   (optional): Set to 1 if HTTPs is implemented on Grafana.
+                                     defaults to 0 (false)
+        GRAFANA_API_TOKEN          : Grafana API authentication token
+    """
+
+    CALLBACK_VERSION = 1.0
+    CALLBACK_TYPE = 'aggregate'
+    CALLBACK_NAME = 'grafana'
+    CALLBACK_NEEDS_WHITELIST = True
+
+    def __init__(self):
+        super(CallbackModule, self).__init__()
+
+        token = os.getenv('GRAFANA_API_TOKEN')
+        self.secure = int(os.getenv('GRAFANA_SECURE', 0))
+        if self.secure not in [0, 1]:
+            self.secure = 0
+        if token is None:
+            self.disabled = True
+            self._display.warning("GRAFANA_API_TOKEN not defined in the environment")
+        self.grafana_server = os.getenv('GRAFANA_SERVER', 'localhost')
+        self.grafana_port = int(os.getenv('GRAFANA_PORT', 3000))
```

Comment:
why not use only one variable `grafana_url` like https://grafana-server:3000 instead of 3 distinct variable host / port and secure

Reply:
I'll repost with only 2, merging grafana_server and grafana_port. I'll keep the secure one mainly to avoid parsing the string looking for http or https.